src/tokenizedSubaccounts/PPTSA.sol

```diff
@@ -0,0 +1,597 @@
+        || pptsaParams.maxNegCash > 0 || pptsaParams.rfqFeeFactor > 0.1e18 || pptsaParams.maxLossOrGainPercentOfTVL > 1e18
+        || pptsaParams.negMaxCashTolerance < 1e16 || pptsaParams.negMaxCashTolerance > 1e18
+        || ($.isLongSpread && pptsaParams.maxTotalCostTolerance < 1e18)
+        || (!$.isLongSpread && pptsaParams.maxTotalCostTolerance > 1e18)
+    ) {
+      revert PPT_InvalidParams();
+    }
+    _getPPTSAStorage().ppParams = pptsaParams;
+    _getPPTSAStorage().collateralManagementParams = newCollateralMgmtParams;
+
+    emit PPTSAParamsSet(pptsaParams, newCollateralMgmtParams);
+  }
+
+  function _getCollateralManagementParams() internal view override returns (CollateralManagementParams storage $) {
+    return _getPPTSAStorage().collateralManagementParams;
+  }
+
+  ///////////////////////
+  // Action Validation //
+  ///////////////////////
+  function _verifyAction(IMatching.Action memory action, bytes32 actionHash, bytes memory extraData)
+    internal
+    virtual
+    override
+  {
+    PPTSAStorage storage $ = _getPPTSAStorage();
+
+    if (
+      action.expiry < block.timestamp + $.ppParams.minSignatureExpiry
+        || action.expiry > block.timestamp + $.ppParams.maxSignatureExpiry
+    ) {
+      revert PPT_InvalidActionExpiry();
+    }
+
+    _revokeSignature($.lastSeenHash);
+    $.lastSeenHash = actionHash;
+
+    BaseTSAAddresses memory tsaAddresses = getBaseTSAAddresses();
+
+    if (address(action.module) == address($.depositModule)) {
+      _verifyDepositAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.withdrawalModule)) {
+      _verifyWithdrawAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.tradeModule)) {
+      _verifyTradeAction(action, tsaAddresses);
+    } else if (address(action.module) == address($.rfqModule)) {
+      _verifyRfqAction(action, extraData);
+    } else {
+      revert PPT_InvalidModule();
+    }
+  }
+
+  /////////////////
+  // Withdrawals //
+  /////////////////
+
+  function _verifyWithdrawAction(IMatching.Action memory action, BaseTSAAddresses memory tsaAddresses) internal view {
+    IWithdrawalModule.WithdrawalData memory withdrawalData = abi.decode(action.data, (IWithdrawalModule.WithdrawalData));
+
+    if (withdrawalData.asset != address(tsaAddresses.wrappedDepositAsset)) {
+      revert PPT_InvalidAsset();
+    }
+
+    (uint openSpreads, uint baseBalance, int cashBalance) = _getSubAccountStats();
+
+    if (openSpreads != 0) {
+      revert PPT_WithdrawingWithOpenTrades();
+    }
+
+    uint withdrawalAs18Decimals =
+      ConvertDecimals.to18Decimals(withdrawalData.assetAmount, tsaAddresses.depositAsset.decimals());
+    if (withdrawalAs18Decimals > baseBalance) {
+      revert PPT_InvalidBaseBalance();
+    }
+
+    if (cashBalance >= 0) {
+      return;
+    }
+
+    uint remainingBaseBalance = baseBalance - withdrawalAs18Decimals;
```

Comment:
nit: Would maybe add (...).multiplyDecimal(_getBasePrice()) here instead of in the if below; maybe that keeps it to one line

Reply:
Added